substrate/frame/revive/src/primitives.rs

```diff
@@ -76,19 +76,63 @@ pub struct ContractResult<R, Balance, EventRecord> {
 	/// RPC calls.
 	pub debug_message: Vec<u8>,
 	/// The execution result of the wasm code.
-	pub result: R,
+	pub result: Result<R, DispatchError>,
 	/// The events that were emitted during execution. It is an option as event collection is
 	/// optional.
 	pub events: Option<Vec<EventRecord>>,
 }
 
-/// Result type of a `bare_call` call as well as `ContractsApi::call`.
-pub type ContractExecResult<Balance, EventRecord> =
-	ContractResult<Result<ExecReturnValue, DispatchError>, Balance, EventRecord>;
+/// The result of the execution of a `eth_transact` call.
+pub struct EthContractResultDetails<Balance> {
+	/// The kind of transaction that was executed.
+	pub transact_kind: EthTransactKind,
+	/// The call's dispatch info.
+	pub dispatch_info: DispatchInfo,
+	/// Length of the encoded call.
+	pub len: u32,
+	/// Gas limit of the transaction.
+	pub gas_limit: Weight,
+	/// Storage deposit charged.
+	pub storage_deposit: Balance,
+	/// The execution result.
+	pub result: Result<Vec<u8>, DispatchError>,
+}
 
-/// Result type of a `bare_instantiate` call as well as `ContractsApi::instantiate`.
-pub type ContractInstantiateResult<Balance, EventRecord> =
-	ContractResult<Result<InstantiateReturnValue, DispatchError>, Balance, EventRecord>;
+impl<Balance: From<u32>> EthContractResultDetails<Balance> {
+	/// Map to a EthContractResult, using the provided compute_fee function.
+	///
+	/// # Parameters
+	///
+	/// - `compute_fee`: A function that takes the length of the encoded call, the dispatch info and
+	///   the tip and returns the fee.
+	pub fn map(
+		self,
+		compute_fee: impl FnOnce(u32, &DispatchInfo, Balance) -> Balance,
+	) -> EthContractResult<Balance> {
+		EthContractResult {
+			transact_kind: self.transact_kind,
+			result: self.result,
+			gas_limit: self.gas_limit,
+			storage_deposit: self.storage_deposit,
+			fee: compute_fee(self.len, &self.dispatch_info, 0.into()),
+		}
+	}
+}
```

Comment:
I don't think we need this. We are already tightly coupled with `pallet-transaction-payment` and can just calculate the fee inside `bare_eth_transact`.  No need for the details type.